stan/math/prim/mat/fun/gp_exp_quad_cov.hpp

```diff
@@ -240,6 +246,300 @@ gp_exp_quad_cov(const std::vector<Eigen::Matrix<T_x1, Eigen::Dynamic, 1>> &x1,
   }
   return cov;
 }
+
+#ifdef STAN_OPENCL
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @param x std::vector of scalars that can be used in square distance.
+ * @param sigma marginal standard deviation or magnitude
+ * @param length_scale length scale
+ * @return squared distance
+ * @throw std::domain_error if sigma <= 0, l <= 0, or
+ *   x is nan or infinite
+ */
+template <typename T = void>  // if this was non-templated overload or fully
+// specialized template, the compiler could
+// incorrectly resolve some called functions to
+// prim implementation instead of arr
+inline Eigen::MatrixXd gp_exp_quad_cov(const std::vector<double> &x,
+                                       const double sigma,
+                                       const double length_scale) {
+  const char *function_name = "gp_exp_quad_cov";
+  check_positive(function_name, "magnitude", sigma);
+  check_positive(function_name, "length scale", length_scale);
+
+  const size_t x_size = x.size();
+  Eigen::MatrixXd cov(x_size, x_size);
+  if (x_size * x_size < opencl_context.tuning_opts().gp_exp_quad_cov_size) {
+    // using explicit template args to call CPU function
+    return gp_exp_quad_cov<double, double, double>(x, sigma, length_scale);
+  }
+
+  if (x_size == 0)
+    return cov;
+
+  matrix_cl x_cl(x, 1, x.size());
+  check_nan(function_name, "x", x_cl);
+  matrix_cl cov_cl = gp_exp_quad_cov(x_cl, sigma, length_scale);
+  copy(cov, cov_cl);  // NOLINT
+
+  return cov;
+}
+
+/**
+ * Returns a squared exponential kernel.
+ *
+ * @param x std::vector of scalars that can be used in square distance.
+ *    This function assumes each element of x is the same size.
+ * @param sigma marginal standard deviation or magnitude
+ * @param length_scale length scale
+ * @return squared distance
+ * @throw std::domain_error if sigma <= 0, l <= 0, or
+ *   x is nan or infinite
+ */
+template <typename T = void>  // if this was non-templated overload or fully
+// specialized template, the compiler could
+// incorrectly resolve some called functions to
+// prim implementation instead of arr
+inline Eigen::MatrixXd gp_exp_quad_cov(const std::vector<Eigen::VectorXd> &x,
+                                       const double sigma,
+                                       const double length_scale) {
+  const char *function_name = "gp_exp_quad_cov";
+  check_positive(function_name, "magnitude", sigma);
+  check_positive(function_name, "length scale", length_scale);
+
+  const size_t x_size = x.size();
+  Eigen::MatrixXd cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  const size_t s = x[0].size();
+  if (x_size * x_size * opencl_context.tuning_opts().gp_exp_quad_cov_coeff1
+          + (x_size + x_size + 1) * s
+                * opencl_context.tuning_opts().gp_exp_quad_cov_coeff2
+      < 1) {
```

Comment:
One part of work done on either CPU of GPU is proportional to input data size and another part is proportional to output data size. Output size equals `x_size * x_size` or `x1_size * x2_size` for the case with two input vectors. Input size equals `(x_size + x_size + 1) * s` or `(x1_size + x2_size + 1) * s` for the case with two input vectors.

`gp_exp_quad_cov_coeff1` and `gp_exp_quad_cov_coeff2` are weights how much these parts affect GPU speedup. I fitted these to coefficcients to measurements I earlier posted on this thread.

Reply:
Can we make the values for `gp_exp_quad_cov_coeff1` and `gp_exp_quad_cov_coeff2` real values instead of proportions? It's harder to reason about them when they are proportions of the inputs. I'm building at test right now that will just check given the size of N for the different setups how the OpenCL and Eigen stuff differs

Reply:
I started a folder we can use in a branch in perf-math below

https://github.com/seantalts/perf-math/blob/gp_exp_quad_speed_tests/tests/gp_exp_quad_cov/gp_exp_quad_cov_speed.cpp

Reply:
I have to hit the hay but can lot at this more tomorrow

Reply:
I missed this comment ... 
What do you mean with "real values instead of proportions"? Do you mean integers instead of floats or something else?

Reply:
integer values. It would make it easier to maintain and reason about.

I just saw your benchmark code above. Can you rerun these all for the same values, say starting at 700 and then increment by 200 or so? I would but I only have access to my laptop at this moment

https://gist.github.com/t4c1/f7a8606c03e03f105f53fc51f169f757

Reply:
I will replace it with `1/current value`, which can be integer. Ok, I will rerun these tests. Actually I think I can modify them a bit to better show that my choices of values are good. But it is getting late here, so I will do that tomorrow.